### DISCUSSION TITLE
How to do Views or derived state

### DISCUSSION THREAD

**Discussion by codingedgar** [statelyai/xstate#2221] [Q&A]
Dear #xstate folks, what are some approaches to building “views” out of the context/state?

I’m often torn between “making the context minimal” and “not transforming data at the UI level”.

I stopped using the context as a full view because it becomes too hard to update things, mentally, I often leave data out of sync.

It is easier to have single sources of thrust like a map/array with elements and indexes.

But then I’ll have to map/filter/reduce data on the components, often perform the same transforms multiple times.

So I’m not saving time nor space, I think(?), it would be the same if I just have a view somewhere that gets computed every time the state changes? Feels slow perf.

original tweet: https://twitter.com/codingedgar/status/1395425468115951617

### Update:

Thanks to @davidkpiano suggestions what I've been using (which is pretty much ok for react) is the following:

```
const service = useInterpret(machine):
const { send } = service;
const value = useSelector(service, selectValue);
```

And so far has been working quite well, with much fewer re-renders and improved readability for the actual JSX.

**Comment by davidkpiano** [comment]
Can you show a code example of what you're currently doing? I've been wanting to support derived (computed) views; just need to get the API right.

**Reply by codingedgar** [reply]
My use case is pretty much the same as @with-heart, I often do:

``` ts
derivedValues: {
	  // sets `context.userName`
      userName: (context) => {
        const {firstName, lastName} = context.user
        return `${firstName} ${lastName}`
      },
	  // sets `context.userAddress`
      userAddress: (context) => {
        const {userName, user} = context
        const {street, city, state, zip} = user.location
        const address = `${userName} ${street} ${city}, ${state} ${zip}`.trim()
        return address
      },
```

A difference not stated in the example is that I often do use `state.matches` to make booleans to show/hide components, I rarely use the state to derive data differently itself, but the efforts to make the Context more "state aware" for sure I'll lay on that more, the context IS susceptible to state in my own experience, but i often just use `fp-ts`'s `Option` to deal with that so the transforms state-unaware.

**Comment by ChrisShank** [comment]
Just linking a previous discussion for reference! https://github.com/davidkpiano/xstate/discussions/2111

**Comment by with-heart** [comment]
I've been thinking about this off and on the last week or two as I've run into various situations where I derived variables/state in `React` components from `context`.

I don't know how I feel about the ergonomics of this example, but it's what has been floating around in my mind and I wanted to just get it out:

```ts
import {createMachine, interpret} from 'xstate'
import {createModel} from 'xstate/lib/model'

interface User {
  firstName: string
  lastName: string
  currentLocation: Location
}

interface Location {
  street: string
  city: string
  state: string
  zip: string
}

const model = createModel(
  {
    user: {
      firstName: 'Mark',
      lastName: 'Chandler',
      currentLocation: {
        street: '123 City Ln',
        city: 'Blue Ash',
        state: 'Ohio',
        zip: '45242',
      },
    } as User,
    locations: [
      {
        street: '123 City Ln',
        city: 'Blue Ash',
        state: 'Ohio',
        zip: '45242',
      },
      {
        street: '456 Town Blvd',
        city: 'Saint Bernard',
        state: 'Ohio',
        zip: '45217',
      },
      {
        street: '789 Village Ave',
        city: 'Reading',
        state: 'Ohio',
        zip: '45202',
      },
    ] as Location[],
  },
  {
    events: {
      setLocation: (location: Location) => ({location}),
    },
  },
)

const machine = createMachine<typeof model>(
  {
    context: model.initialContext,
    initial: 'state',
    states: {
      state: {
        on: {setLocation: {actions: 'setLocation '}},
      },
    },
  },
  {
    actions: {
      setLocation: model.assign(
        {
          user: (context, event) => ({
            ...context.user,
            currentLocation: event.location,
          }),
        },
        'setLocation',
      ),
    },
	// values derived when context changes
    derivedValues: {
	  // sets `context.userName`
      userName: (context) => {
        const {firstName, lastName} = context.user
        return `${firstName} ${lastName}`
      },
	  // sets `context.userAddress`
      userAddress: (context) => {
        const {userName, user} = context
        const {street, city, state, zip} = user.location
        const address = `
${userName}
${street}
${city}, ${state} ${zip}
        `.trim()
        return address
      },
    },
  },
)

const service = interpret(machine).onTransition((state) => {
  console.log(state.context.userAddress)
})

service.start()

/*
Mark Chandler
123 City Ln
Blue Ash, Ohio 45242
*/

service.send(
  model.events.setLocation({
    street: '456 Town Blvd',
    city: 'Saint Bernard',
    state: 'Ohio',
    zip: '45217',
  }),
)

/*
Mark Chandler
456 Town Blvd
Saint Bernard, Ohio 45217
*/
```

The idea is that the keys of `derivedValues` correspond to the `context` key under which the derived value is stored and the values are the functions that specify how that value is derived. The functions are called when `context` changes (post-`event` resolution).

Possible positives:

- Much easier to define derivations once rather than having to call them as actions all the time
- `derivedValues` follows a familiar interface for selecting things from `context`
- `derivedValues` can be overridden via `.withConfig`, allowing the consumer to customize how values are derived without modifying anything else about `machine`
- Depending on how this worked with models, we could have a `model.derive` convenience function similar to `model.assign`

Some questions that came to mind as I wrote the example:

- Is it okay if a derived value uses the same key of a non-derived value on `context`?
- Does it even make sense for derived values to live in `context`? It's nice have a unified interface for accessing data in the machine, but I could see this being confusing.
- How would we type derived values?
- Can we predictably handle derived values that depend on other derived values? For example, `userAddress` depends on `userName`, so `userName` would need to be derived first.

**Reply by codingedgar** [reply]
Thanks for the `createModel` use case, wasn't aware it was a thing.

My use case is pretty much the same thing, my two cents:

- I don't think `derivedValues` should live in the context, I think a more declarative property could be named "views" and that would pretty much sum up what it should do, a view, as many other systems, a rendered version of the context+state+etc. `(Context, Event, Meta) => View`

- I think solving dependencies between the properties of the own views is probably one of those no-nos, It could be just like the Context, if you want consistency between two properties, you build them in the same step.

**Comment by davidkpiano** [comment]
Try this:

```jsx
function compute(views, state) {
  const computed = {};
  const stack = [];
  const viewProxy = new Proxy(views, {
    get: (target, prop) => {
      if (computed[prop]) {
        return computed[prop];
      }
      if (stack.includes(prop)) {
        throw new Error(`Cyclic views: ${stack.join(" → ")} → ${prop}`);
      }
      stack.push(prop);
      computed[prop] = views[prop](state, viewProxy);
      stack.pop();
      return computed[prop];
    }
  });

  return viewProxy;
}

const views = {
  count: (state) => state.context.count,
  times: (_, views) => (views.count === 1 ? "time" : "times"),
  countText: (state, views) => (
    <span>
      Toggled <strong>{views.count}</strong> {views.times}
    </span>
  )
};

const { count, times, countText } = compute(
  views,
  currentState
);
```

**Reply by codingedgar** [reply]
Ok, gonna use it and report back how it feels.

I was experimenting with mobx which uses the same principle (proxy), but was reluctant to add a third state management lib.

The idea of computing/catching the JSX in the views is 👌
